README
===========
WHAT IS IT?
===========

This package provides the console with the same keyboard configuration
scheme that X Window System has. In result, there is no need to
duplicate or change the console keyboard files just to make simple
customizations such as the use of dead keys, the key functioning as
AltGr or Compose key, the key(s) to switch between Latin and non-Latin
layouts, etc.  Besides the keyboard, the package configures also the
font on the console.  It includes a rich collection of fonts and
supports several languages that would be otherwise unsupported on the
console (such as Armenian, Georgian, Lao and Thai).

The package supports Linux and FreeBSD kernels.  On Linux it supports
PC, Amiga, Atari, old Macintosh, Sun4 and Sun5 keyboards.  On FreeBSD
it supports PC keyboards.

There are no binaries in the package: it can be compiled on one
platform and used on another.  For its compilation it requires GNUmake
and Perl.  

The package can be installed in two ways -- with or without
precompiled keyboard files.  If it is installed with precompiled
keyboard files, then it won't require Perl for its normal work, but
not all possible keyboard configurations will be properly supported.
If the package is installed without precompiled keyboard files, then
it will require Perl for its normal work but you will be able to use
arbitrary keyboard configuration on the console.


============
INSTALLATION
============

First you need to compile the package.  Notice, that this can take a
long time, especially if you will use precompiled keyboard files.  Use
one of the following commands:

   make build-linux         # For Linux without precompiled keyboard files
   make build-mini-linux    # For Linux with precompiled keyboard files
   gmake build-freebsd      # For FreeBSD without precompiled keyboard files
   gmake build-mini-freebsd # For FreeBSD with precompiled keyboard files

Then in order to install the package in /usr/local use one of the
following commands:

   make install-linux
   make install-mini-linux
   gmake install-freebsd
   gmake install-mini-freebsd

You don't have to install the package in /usr/local -- you can install
it anywhere and test how it works from there, for example:

   mkdir /tmp/cs
   make prefix=/tmp/cs install-linux 

The Makefile also provides uninstall targets (such as uninstall-linux)
but if you have installed 'stow', then it is recommended to use it for
the deinstallation.

In order to use console-setup it is not necessary to have installed X
Window because the package comes with its own version of the X
keyboard files.  By default they are installed in
/usr/local/etc/console-setup/ckb.  If, however, X Window is installed,
then you can remove this directory and console-setup will use the same
files X uses.

If X Window is installed in the system and you want to use
console-setup with precompiled keyboard files, then compile it with

   make xkbdir=/usr/local/share/X11/xkb build-mini-linux


===========
BASIC USAGE
===========

First look at the configuration files keyboard and console-setup.  By
default they will be installed in /usr/local/etc/default.  Read their
manual pages: keyboard(5) and console-setup(5).  You can find
preformatted text version of these manual pages in the directory man/
of the source package.

When you are ready, run the command 'setupcon' without parameters and
that's it.  In order to make your system automatically configure the
console, simply put the command 'setupcon' in some of the boot
scripts.

If you are using FreeBSD, create a file /etc/rc.local with the
following contents:

------------------------
#!/bin/sh

setupcon
------------------------

Then create a file /etc/rc.conf.local with the following contents
(this will make console-setup the only package responsible to
configure the console):

------------------------
font8x8="NO"
font8x14="NO"
font8x16="NO"
scrnmap="NO"
keymap="NO"
------------------------


============================
TERMINAL TYPE (FreeBSD only)
============================

In order to display properly the line draw symbols on FreeBSD, the
programs use the description of the capabilities of the terminal of
the console in order to find the position of these symbols in the
current encoding.  Traditionally this meant that the different
encodings ought to be used with different terminal types.  Obviously
this is very inconvenient for a multi-lingual package such as
console-setup.  In order to overcome this difficulty, for the line draw
symbols console-setup uses some of the special codes from 1 to 31.  In
this way only one unified terminal type can be used for all languages.

Console-setup provides the description of the terminal it uses in two
formats -- the older termcap and the newer terminfo.  They are
installed in /usr/local/etc/console-setup/.

On FreeBSD the older format is used (termcap).  You have to append the
console-setup definition to the system-wide file and to recompile it:

   cat /usr/local/etc/console-setup/termcap >>/etc/termcap
   cap_mkdb -f /usr/share/misc/termcap /etc/termcap

Then open /etc/ttys in a text editor and find all lines whose first
word is of the form ttyv0, ttyv1, etc:

   ttyv0   "/usr/libexec/getty Pc"         cons25  on  secure

Replace cons25 with cons25cs (cs is an abbreviation of console-setup).
Console-setup provides also the following alternative terminal types:
cons25cs-m, cons30cs, cons30cs-m, cons43cs, cons43cs-m, cons50cs,
cons50cs-m, cons60cs and cons60cs-m.  The number tells how many lines
are used on the console and the suffix -m means 'monochrome'.  If for
example the console is configured to use video mode 80x30 with color
video display, then in /etc/ttys use cons30cs instead of cons25cs

If you login from the console to a remote system that doesn't
understands the terminal types of console-setup, this can be a
problem.  In case the remote system uses the newer terminfo format
(this is true on Linux systems) you can solve the problem in the
following way (there is no need to be root on the remote system):

   local:  scp /usr/local/etc/console-setup/terminfo remote.system.org:.
   local:  ssh remote.system.org
   remote: tic terminfo
   remote: rm terminfo

I don't know if something similar can be done as non-root user on
systems that use termcap.  If you login in a such system, then change
the terminal type to 'cons25w':

   TERM=cons25w; export TERM   # if you use Bourne-alike shell
   setenv TERM cons25w         # if you use the C shell


====================
BACKSPACE AND DELETE
====================

This hasn't been a problem on Linux since a while.  But on FreeBSD it
is not unusual to have programs that do not respond correctly to these
keys.  For example in some situations Backspace will delete the symbol
at the cursor (not the preceding symbol) in other situations the
Backspace key will display help information (because this key actually
generates the combination Control-h).

I don't know why proper solution of the problem hasn't been applied by
the FreeBSD project, considering how easy is it to do so.  After I
reconfigured properly the keys Backspace and Delete, all programs I
tested worked correctly (absolutely no reconfiguration of these
programs was necessary).

In general, there are two standards for the codes these keys generate.
The first one strives to be conformant with the behavior of the old
DEC VT100 terminal.  This standard is used by default on FreeBSD.  The
second standard is conformant with the newer terminal VT220.  The
following table summarizes the codes generated by the keys according
to these two standards (^H is BS, ^? is DEL and ^[ is ESC):

              VT100  VT220
   Control-h   ^H     ^H
   Backspace   ^H     ^?
   Delete      ^?    ^[[3~

Notice that VT100 makes Backspace and Control-h to behave identically.
This means the programs (notably Emacs) are unable to differentiate
between Backspace and Control-h.  No reconfiguration can help this.
The problem is aggravated by the fact that the other standard (that of
VT220) is widespread on Linux systems, whence many programs require
reconfiguration in order to work correctly with VT100 codes.

There are two methods the programs can use in order to determine the
properties of the terminal they are working on and in particular the
codes generated by the keys Backspace and Delete.  The first method is
to read the settings of the terminal line.  This method gives limited
information; it is, however, in a sense directly supported by the
kernel of the operating system.  Use the command 'stty -a' to see
these settings.

The second method is based on the variable of the environment TERM.
Use 'echo $TERM' to see its value.  This variable tells the programs
the name of the terminal you are using.  When the name of the terminal
is known, the programs look in a database describing the capabilities
of all terminals in order to determine the capabilities of your
particular terminal.  There are two implementations of such a database
-- termcap and the terminfo, but for our purposes it makes no
difference which one is used on your system.

Somewhat inconsistently, two different programs are responsible for
the configuration of the terminal capabilities -- init and getty:

1. Init is the first process.  It configures the environment and in
   particular it is responsible for the value of the variable TERM.

2. Init starts several instances of getty.  Each virtual terminal has

debian/changelog
console-setup (1.205) unstable; urgency=medium

  * Team upload.

  [ Updated translations ]
  * Finnish (fi.po) by Kimmo Kujansuu

 -- Holger Wansing <hwansing@mailbox.org>  Fri, 23 Jul 2021 19:05:53 +0200

console-setup (1.204) unstable; urgency=medium

  * Team upload.

  [ Updated translations ]
  * Lithuanian (lt.po) by Gediminas Murauskas
  * Turkish (tr.po) by Fatih Altun

 -- Holger Wansing <hwansing@mailbox.org>  Thu, 08 Jul 2021 16:30:43 +0200

console-setup (1.203) unstable; urgency=medium

  * Team upload.

  [ Updated translations ]
  * Arabic (ar.po) by Fahim Sabah
  * Greek (el.po) by Michalis
  * Persian (fa.po) by Sina Carnelious
  * Japanese (ja.po) by Nozomu KURASAWA
  * Polish (pl.po) by Matthaiks

 -- Holger Wansing <hwansing@mailbox.org>  Sun, 30 May 2021 22:30:16 +0200

console-setup (1.202) unstable; urgency=medium

  * Team upload.

  [ Updated translations ]
  * Arabic (ar.po) by zer0-x
  * Hindi (hi.po) by Aryan L
  * Lithuanian (lt.po) by Kornelijus Tvarijanavičius

 -- Holger Wansing <hwansing@mailbox.org>  Sat, 20 Mar 2021 21:54:30 +0100

console-setup (1.201) unstable; urgency=medium

  * Team upload.

  [ Updated translations ]
  * Arabic (ar.po) by zer0-x
  * Bengali (bn.po) by Indrani Roy
  * Catalan (ca.po) by d
  * Danish (da.po) by Michael Millet
  * Spanish (es.po) by Javier FernÃ¡ndez-Sanguino
  * Estonian (et.po) by Kristjan RÃ¤ts
  * Galician (gl.po) by mantinan
  * Hindi (hi.po) by KushagraKarira
  * Kabyle (kab.po) by Selyan Sliman Amiri
  * Latvian (lv.po) by Tranzistors
  * Tamil (ta.po) by Vasudevan Tirumurti
  * Thai (th.po) by Theppitak Karoonboonyanan
  * Simplified Chinese (zh_CN.po) by TTT DEAL
  * Traditional Chinese (zh_TW.po) by louies0623

 -- Holger Wansing <hwansing@mailbox.org>  Thu, 18 Feb 2021 22:42:35 +0100

console-setup (1.200) unstable; urgency=medium

  * Team upload.

  * Fix missing changelog entries for 1.199.

  [ Updated translations ]
  * Icelandic (is.po) by Sveinn Ã Felli
  * Romanian (ro.po) by Sergiu
  * Serbian (sr.po) by Filipovic Dragan

 -- Holger Wansing <hwansing@mailbox.org>  Thu, 31 Dec 2020 22:19:38 +0100

console-setup (1.199) unstable; urgency=medium

  * Team upload.

  [ Holger Wansing ]
  * Remove duplicated Section: and Priority: fields.

  [ Updated translations ]
  * Bulgarian (bg.po) by Damyan Ivanov
  * Czech (cs.po) by Miroslav Kure
  * German (de.po) by Holger Wansing
  * Greek (el.po) by EG
  * Esperanto (eo.po) by Felipe Castro
  * Basque (eu.po) by Iñaki Larrañaga Murgoitio
  * French (fr.po) by Baptiste Jammet
  * Hebrew (he.po) by Yaron Shahrabani
  * Croatian (hr.po) by Milo Ivir
  * Indonesian (id.po) by Andika Triwidada
  * Kabyle (kab.po) by Slimane Selyan Amiri
  * Kurdish (ku.po) by Lecwan Munzur
  * Norwegian Bokmal (nb.po) by Allan Nordhøy
  * Dutch (nl.po) by Frans Spiesschaert
  * Occitan (oc.po) by Quentin PAGÈS
  * Punjabi (Gurmukhi) (pa.po) by Aman Alam
  * Portuguese (pt.po) by Miguel Figueiredo
  * Brazilian Portuguese (pt_BR.po) by Adriano Rafael Gomes
  * Russian (ru.po) by Yuri Kozlov
  * Swedish (sv.po) by Kristoffer Grundström

 -- Cyril Brulebois <kibi@debian.org>  Tue, 01 Dec 2020 02:02:30 +0100

console-setup (1.198) unstable; urgency=medium

  * Team upload

  * Add Berber (Latin) keymap for new language Kabyle.

  [ Updated translations ]
  * Bulgarian (bg.po) by Damyan Ivanov
  * Czech (cs.po) by Miroslav Kure
  * Persian (fa.po) by سهیل خانعلی‌پور
  * Italian (it.po) by Milo Casagrande
  * Marathi (mr.po) by Prachi Joshi
  * Norwegian Bokmal (nb.po) by Petter Reinholdtsen
  * Punjabi (Gurmukhi) (pa.po) by Aman ALam
  * Traditional Chinese (zh_TW.po) by Walter Cheuk

  [ New translations ]
  * Kabyle (kab.po) by Slimane Selyan Amiri
  * Occitan (oc.po) by Quentin PAGÈS

 -- Holger Wansing <hwansing@mailbox.org>  Thu, 29 Oct 2020 18:59:47 +0100

console-setup (1.197) unstable; urgency=medium

  * Team upload

  [ Anton Zinoviev ]
  * standard.equivalents: map more aggressively the box drawing symbols.
    Mappings for ballot symbols.  Thanks to Nick Black, closes: #965029.
  * sync the Terminus font with the xfonts-terminus package (new upstream
    version with new font size 18x10, apply the patches dv1.diff (modified
    Cyrillic letters d and v) and td1.diff (centered tilde)).
  * console-setup.templates: remove references for the console-tools
    package which is no longer in the Debian archive.  Do not refer to the
    font size at boot time which is irrelevant with framebuffer.

  [ Updated translations ]
  * German (de.po) by Holger Wansing
  * Greek (el.po) by root
  * Esperanto (eo.po) by Felipe Castro
  * Spanish (es.po) by Javier Fernández-Sanguino
  * Estonian (et.po) by Kristjan Räts

Keyboard/KeyboardNames.pl
#!/usr/bin/perl -w

package KeyboardNames;

%models = (
    'A4Tech KB-21' => 'a4techKB21',
    'A4Tech KBS-8' => 'a4techKBS8',
    'ACPI Standard' => 'acpi',
    'Acer AirKey V' => 'airkey',
    'Acer Ferrari 4000' => 'acer_ferrari4k',
    'Acer TravelMate 800' => 'acer_tm_800',
    'Advance Scorpius KI' => 'scorpius',
    'Azona RF2300 wireless Internet Keyboard' => 'azonaRF2300',
    'BTC 5090' => 'btc5090',
    'BTC 5113RF Multimedia' => 'btc5113rf',
    'BTC 5126T' => 'btc5126t',
    'BTC 9000' => 'btc9000',
    'BTC 9000A' => 'btc9000a',
    'BTC 9001AH' => 'btc9001ah',
    'Brazilian ABNT2' => 'abnt2',
    'Brother Internet Keyboard' => 'brother',
    'Cherry Blue Line CyBo@rd' => 'cherryblue',
    'Cherry Blue Line CyBo@rd (alternate option)' => 'cherrybluea',
    'Cherry CyBo@rd USB-Hub' => 'cherrycyboard',
    'Cherry CyMotion Master Linux' => 'cymotionlinux',
    'Cherry CyMotion Master XPress' => 'cherryblueb',
    'Chicony Internet Keyboard' => 'chicony',
    'Chicony KB-9885' => 'chicony9885',
    'Compaq Easy Access Keyboard' => 'compaqeak8',
    'Compaq Internet Keyboard (13 keys)' => 'compaqik13',
    'Compaq Internet Keyboard (18 keys)' => 'compaqik18',
    'Compaq Internet Keyboard (7 keys)' => 'compaqik7',
    'Compaq iPaq Keyboard' => 'ipaq',
    'DTK2000' => 'dtk2000',
    'Dell' => 'dell',
    'Dell 101-key PC' => 'dell101',
    'Dell USB Multimedia Keybard' => 'dellusbmm',
    'Dexxa Wireless Desktop Keyboard' => 'dexxa',
    'Diamond 9801 / 9802 series' => 'diamond',
    'Ennyah DKB-1008' => 'ennyah_dkb1008',
    'Everex STEPnote' => 'everex',
    'Generic 101-key PC' => 'pc101',
    'Generic 102-key (Intl) PC' => 'pc102',
    'Generic 104-key PC' => 'pc104',
    'Generic 105-key (Intl) PC' => 'pc105',
    'Genius Comfy KB-12e' => 'geniuscomfy',
    'Genius Comfy KB-16M / Genius MM Keyboard KWD-910' => 'genius',
    'Genius Comfy KB-21e-Scroll' => 'geniuscomfy2',
    'Gyration' => 'gyration',
    'Hewlett-Packard Internet Keyboard' => 'hpi6',
    'Hewlett-Packard Omnibook 500 FA' => 'hp500fa',
    'Hewlett-Packard Omnibook 5xx' => 'hp5xx',
    'Hewlett-Packard Omnibook 6000/6100' => 'hp6000',
    'Hewlett-Packard Omnibook XE3 GC' => 'hpxe3gc',
    'Hewlett-Packard Omnibook XE3 GF' => 'hpxe3gf',
    'Hewlett-Packard Omnibook XT1000' => 'hpxt1000',
    'Hewlett-Packard Pavilion ZT11xx' => 'hpzt11xx',
    'Hewlett-Packard SK-2501 Multimedia Keyboard' => 'hp2501',
    'Hewlett-Packard SK-2505 Internet Keyboard' => 'hp2505',
    'Honeywell Euroboard' => 'honeywell_euroboard',
    'IBM Rapid Access' => 'rapidaccess',
    'IBM Rapid Access II' => 'rapidaccess2',
    'IBM Rapid Access II (alternate option)' => 'rapidaccess2a',
    'IBM ThinkPad 560Z/600/600E/A22E' => 'thinkpad',
    'Japanese 106-key' => 'jp106',
    'Keytronic FlexPro' => 'flexpro',
    'Laptop/notebook Compaq (eg. Armada) Laptop Keyboard' => 'armada',
    'Laptop/notebook Compaq (eg. Presario) Internet Keyboard' => 'presario',
    'Laptop/notebook Dell Inspiron 6xxx/8xxx' => 'inspiron',
    'Laptop/notebook eMachines m68xx' => 'emachines',
    'Logitech Access Keyboard' => 'logiaccess',
    'Logitech Cordless Desktop' => 'ltcd',
    'Logitech Cordless Desktop Navigator' => 'logicdn',
    'Logitech Cordless Desktop Optical' => 'logicdo',
    'Logitech Cordless Desktop Pro' => 'logicdp',
    'Logitech Cordless Desktop Pro (alternate option)' => 'logicdpa',
    'Logitech Cordless Desktop Pro (alternate option2)' => 'logicdpa2',
    'Logitech Cordless Desktop iTouch' => 'logicdit',
    'Logitech Cordless Freedom/Desktop Navigator' => 'logicfn',
    'Logitech Deluxe Access Keyboard' => 'logidak',
    'Logitech Internet Keyboard' => 'logiik',
    'Logitech Internet Navigator Keyboard' => 'logiink',
    'Logitech Ultra-X Keyboard' => 'logiultrax',
    'Logitech iTouch' => 'itouch',
    'Logitech iTouch Cordless Keyboard (model Y-RB6)' => 'logiitc',
    'Logitech iTouch Internet Navigator Keyboard SE' => 'logiinkse',
    'Logitech iTouch Internet Navigator Keyboard SE (USB)' => 'logiinkseusb',
    'Logitech iTouch keyboard Internet Navigator' => 'itouchin',
    'Macintosh' => 'macintosh',
    'Macintosh Old' => 'macintosh_old',
    'Memorex MX1998' => 'mx1998',
    'Memorex MX2500 EZ-Access Keyboard' => 'mx2500',
    'Memorex MX2750' => 'mx2750',
    'Microsoft Internet Keyboard' => 'microsoftinet',
    'Microsoft Internet Keyboard Pro, Swedish' => 'microsoftprose',
    'Microsoft Natural' => 'microsoft',
    'Microsoft Natural Keyboard Pro / Microsoft Internet Keyboard Pro' => 'microsoftpro',
    'Microsoft Natural Keyboard Pro OEM' => 'microsoftprooem',
    'Microsoft Natural Keyboard Pro USB / Microsoft Internet Keyboard Pro' => 'microsoftprousb',
    'Microsoft Office Keyboard' => 'microsoftoffice',
    'Microsoft Wireless Multimedia Keyboard 1.0A' => 'microsoftmult',
    'Northgate OmniKey 101' => 'omnikey101',
    'Oretec MCK-800 MM/Internet keyboard' => 'oretec',
    'PC-98xx Series' => 'pc98',
    'Power G5' => 'power_g5',
    'PowerPC PS/2' => 'powerpcps2',
    'Propeller Voyager (KTEZ-1000)' => 'propeller',
    'QTronix Scorpius 98N+' => 'qtronix',
    'SILVERCREST Multimedia Wireless Keyboard' => 'silvercrest',
    'SK-1300' => 'sk1300',
    'SK-2500' => 'sk2500',
    'SK-6200' => 'sk6200',
    'SK-7100' => 'sk7100',
    'SVEN Ergonomic 2500' => 'sven',
    'Samsung SDM 4500P' => 'samsung4500',
    'Samsung SDM 4510P' => 'samsung4510',
    'Super Power Multimedia Keyboard' => 'sp_inet',
    'Symplon PaceBook (tablet PC)' => 'symplon',
    'Toshiba Satellite S3000' => 'toshiba_s3000',
    'Trust Direct Access Keyboard' => 'trustda',
    'Trust Wireless Keyboard Classic' => 'trust',
    'Winbook Model XP5' => 'winbook',
    'Yahoo! Internet Keyboard' => 'yahoo',
);

%layouts = (
    'Afghanistan' => 'af',
    'Albania' => 'al',
    'Andorra' => 'ad',
    'Arabic' => 'ara',
    'Armenia' => 'am',
    'Azerbaijan' => 'az',
    'Bangladesh' => 'bd',
    'Belarus' => 'by',
    'Belgium' => 'be',
    'Bhutan' => 'bt',
    'Bosnia and Herzegovina' => 'ba',
    'Brazil' => 'br',
    'Bulgaria' => 'bg',
    'Cambodia' => 'kh',
    'Canada' => 'ca',
    'Congo, Democratic Republic of the' => 'cd',
    'Croatia' => 'hr',
    'Czechia' => 'cz',
    'Denmark' => 'dk',
    'Esperanto' => 'epo',
    'Estonia' => 'ee',
    'Faroe Islands' => 'fo',
    'Finland' => 'fi',
    'France' => 'fr',
    'Georgia' => 'ge',
    'Germany' => 'de',
    'Ghana' => 'gh',
    'Greece' => 'gr',
    'Hungary' => 'hu',
    'Iceland' => 'is',
    'India' => 'in',
    'Iran' => 'ir',
    'Ireland' => 'ie',
    'Israel' => 'il',
    'Italy' => 'it',
    'Japan' => 'jp',
    'Kazakhstan' => 'kz',
    'Korea, Republic of' => 'kr',
    'Kyrgyzstan' => 'kg',
    'Laos' => 'la',
    'Latin American' => 'latam',
    'Latvia' => 'lv',
    'Lithuania' => 'lt',
    'Macedonia' => 'mk',
    'Maldives' => 'mv',
    'Malta' => 'mt',
    'Maori' => 'mao',
    'Mongolia' => 'mn',
    'Myanmar' => 'mm',
    'Netherlands' => 'nl',
    'Norway' => 'no',
    'PC-98xx Series' => 'nec_vndr/jp',
    'Pakistan' => 'pk',
    'Poland' => 'pl',
    'Portugal' => 'pt',
    'Romania' => 'ro',
    'Russia' => 'ru',
    'Serbia and Montenegro' => 'cs',
    'Slovakia' => 'sk',
    'Slovenia' => 'si',
    'South Africa' => 'za',
    'Spain' => 'es',
    'Sri Lanka' => 'lk',
    'Sweden' => 'se',
    'Switzerland' => 'ch',
    'Syria' => 'sy',
    'Tajikistan' => 'tj',
    'Thailand' => 'th',
    'Turkey' => 'tr',
    'U.S. English' => 'us',
    'Ukraine' => 'ua',
    'United Kingdom' => 'gb',
    'Uzbekistan' => 'uz',
    'Vietnam' => 'vn',
);

%variants = (
    'af' => {
	'Pashto' => 'ps',
	'Southern Uzbek' => 'uz',
    },
    'am' => {
	'Phonetic' => 'phonetic',
    },
    'ara' => {
	'azerty' => 'azerty',
	'azerty/digits' => 'azerty_digits',
	'digits' => 'digits',
	'qwerty' => 'qwerty',
	'qwerty/digits' => 'qwerty_digits',
    },
    'az' => {
	'Cyrillic' => 'cyrillic',
    },
    'ba' => {
	'US keyboard with Bosnian digraphs' => 'unicodeus',
	'US keyboard with Bosnian letters' => 'us',
	'Use Bosnian digraphs' => 'unicode',
	'Use guillemots for quotes' => 'alternatequotes',
    },
    'bd' => {
	'Probhat' => 'probhat',
    },
    'be' => {
	'Eliminate dead keys' => 'nodeadkeys',
	'ISO Alternate' => 'iso-alternate',
	'Sun dead keys' => 'sundeadkeys',
	'Wang model 724 azerty' => 'wang',
    },
    'bg' => {
	'Phonetic' => 'phonetic',
    },
    'br' => {
	'Eliminate dead keys' => 'nodeadkeys',
    },
    'by' => {
	'Latin' => 'latin',
	'Winkeys' => 'winkeys',
    },
    'ca' => {
	'French (legacy)' => 'fr-legacy',
	'French Dvorak' => 'fr-dvorak',
	'Inuktitut' => 'ike',
	'Multilingual' => 'multix',
	'Multilingual, first part' => 'multi',
	'Multilingual, second part' => 'multi-2gr',
    },
    'ch' => {
	'French' => 'fr',
	'French, Sun dead keys' => 'fr_sundeadkeys',
	'French, eliminate dead keys' => 'fr_nodeadkeys',
	'German, Sun dead keys' => 'de_sundeadkeys',
	'German, eliminate dead keys' => 'de_nodeadkeys',
    },
    'cs' => {
	'Latin' => 'latin',
	'Latin Unicode' => 'latinunicode',
	'Latin Unicode qwerty' => 'latinunicodeyz',
	'Latin qwerty' => 'latinyz',
	'Latin with guillemots' => 'latinalternatequotes',
	'With guillemots' => 'alternatequotes',
	'Z and ZHE swapped' => 'yz',
    },
    'cz' => {
	'With <\|> key' => 'bksl',
	'qwerty' => 'qwerty',
	'qwerty, extended Backslash' => 'qwerty_bksl',
    },
    'de' => {
	'Dead acute' => 'deadacute',
	'Dead grave acute' => 'deadgraveacute',
	'Dvorak' => 'dvorak',
	'Eliminate dead keys' => 'nodeadkeys',
	'Romanian keyboard with German letters' => 'ro',
	'Romanian keyboard with German letters, eliminate dead keys' => 'ro_nodeadkeys',
	'Sun dead keys' => 'sundeadkeys',
    },
    'dk' => {
	'Eliminate dead keys' => 'nodeadkeys',
    },
    'ee' => {
	'Dvorak' => 'dvorak',
	'Eliminate dead keys' => 'nodeadkeys',
    },
    'es' => {
	'Catalan variant with middle-dot L' => 'cat',
	'Dvorak' => 'dvorak',
	'Eliminate dead keys' => 'nodeadkeys',
	'Sun dead keys' => 'sundeadkeys',
    },
    'fi' => {
	'Eliminate dead keys' => 'nodeadkeys',
	'Northern Saami' => 'smi',
    },
    'fo' => {
	'Eliminate dead keys' => 'nodeadkeys',
    },
    'fr' => {
	'Alternative' => 'latin9',
	'Alternative, Sun dead keys' => 'latin9_sundeadkeys',
	'Alternative, eliminate dead keys' => 'latin9_nodeadkeys',
	'Dvorak' => 'dvorak',
	'Eliminate dead keys' => 'nodeadkeys',
	'Sun dead keys' => 'sundeadkeys',
    },
    'gb' => {
	'Dvorak' => 'dvorak',
	'International (with dead keys)' => 'intl',
    },
    'ge' => {
	'Russian' => 'ru',
    },
    'gh' => {
	'Akan' => 'akan',
	'Ewe' => 'ewe',
	'Fula' => 'fula',
	'Ga' => 'ga',
	'Hausa' => 'hausa',
	'Multilingual' => 'generic',
    },
    'gr' => {
	'Eliminate dead keys' => 'nodeadkeys',
	'Extended' => 'extended',
	'Polytonic' => 'polytonic',
    },
    'hr' => {
	'US keyboard with Croatian digraphs' => 'unicodeus',
	'US keyboard with Croatian letters' => 'us',
	'Use Croatian digraphs' => 'unicode',
	'Use guillemots for quotes' => 'alternatequotes',
    },
    'hu' => {
	'101/qwerty/comma/Dead keys' => '101_qwerty_comma_dead',
	'101/qwerty/comma/Eliminate dead keys' => '101_qwerty_comma_nodead',
	'101/qwerty/dot/Dead keys' => '101_qwerty_dot_dead',
	'101/qwerty/dot/Eliminate dead keys' => '101_qwerty_dot_nodead',
	'101/qwertz/comma/Dead keys' => '101_qwertz_comma_dead',
	'101/qwertz/comma/Eliminate dead keys' => '101_qwertz_comma_nodead',
	'101/qwertz/dot/Dead keys' => '101_qwertz_dot_dead',
	'101/qwertz/dot/Eliminate dead keys' => '101_qwertz_dot_nodead',
	'102/qwerty/comma/Dead keys' => '102_qwerty_comma_dead',
	'102/qwerty/comma/Eliminate dead keys' => '102_qwerty_comma_nodead',
	'102/qwerty/dot/Dead keys' => '102_qwerty_dot_dead',
	'102/qwerty/dot/Eliminate dead keys' => '102_qwerty_dot_nodead',
	'102/qwertz/comma/Dead keys' => '102_qwertz_comma_dead',
	'102/qwertz/comma/Eliminate dead keys' => '102_qwertz_comma_nodead',
	'102/qwertz/dot/Dead keys' => '102_qwertz_dot_dead',
	'102/qwertz/dot/Eliminate dead keys' => '102_qwertz_dot_nodead',
	'Eliminate dead keys' => 'nodeadkeys',
	'Standard' => 'standard',
	'qwerty' => 'qwerty',
    },
    'ie' => {
	'CloGaelach' => 'CloGaelach',
	'Ogham' => 'ogam',
	'Ogham IS434' => 'ogam_is434',
	'UnicodeExpert' => 'UnicodeExpert',
    },
    'il' => {
	'Phonetic' => 'phonetic',
	'lyx' => 'lyx',
    },
    'in' => {
	'Bengali' => 'ben',
	'Bengali Probhat' => 'ben_probhat',
	'Gujarati' => 'guj',
	'Gurmukhi' => 'guru',
	'Kannada' => 'kan',
	'Malayalam' => 'mal',
	'Oriya' => 'ori',
	'Tamil' => 'tam',
	'Tamil TAB Typewriter' => 'tam_TAB',
	'Tamil TSCII Typewriter' => 'tam_TSCII',
	'Tamil Unicode' => 'tam_unicode',
	'Telugu' => 'tel',
	'Urdu' => 'urd',
    },
    'ir' => {
	'Keypad' => 'keypad',
	'Pro' => 'pro',
	'Pro Keypad' => 'pro_keypad',
    },
    'is' => {
	'Eliminate dead keys' => 'nodeadkeys',
	'Sun dead keys' => 'Sundeadkeys',
    },
    'it' => {
	'Eliminate dead keys' => 'nodeadkeys',
    },
    'kz' => {
	'Kazakh with Russian' => 'kazrus',
	'Russian with Kazakh' => 'ruskaz',
    },
    'latam' => {
	'Eliminate dead keys' => 'nodeadkeys',
	'Sun dead keys' => 'sundeadkeys',
    },
    'lk' => {
	'Tamil TAB Typewriter' => 'tam_TAB',
	'Tamil Unicode' => 'tam_unicode',
    },
    'lt' => {
	'Standard' => 'std',
	'US keyboard with Lithuanian letters' => 'us',
    },
    'lv' => {
	'Apostrophe () variant' => 'apostrophe',
	'F-letter (F) variant' => 'fkey',
	'Tilde (~) variant' => 'tilde',
    },
    'mk' => {
	'Eliminate dead keys' => 'nodeadkeys',
    },
    'mt' => {
	'US keyboard with Maltian letters' => 'us',
    },
    'nl' => {
	'Sun dead keys' => 'sundeadkeys',
    },
    'no' => {
	'Dvorak' => 'dvorak',
	'Eliminate dead keys' => 'nodeadkeys',
	'Northern Saami' => 'smi',
	'Northern Saami, eliminate dead keys' => 'smi_nodeadkeys',
    },
    'pl' => {
	'Dvorak' => 'dvorak',
	'Dvorak, Polish quotes on key "1/!"' => 'dvorak_altquotes',
	'Dvorak, Polish quotes on quotemark key' => 'dvorak_quotes',
	'qwertz' => 'qwertz',
    },
    'pt' => {
	'Eliminate dead keys' => 'nodeadkeys',
	'Sun dead keys' => 'sundeadkeys',
    },
    'ro' => {
	'Standard' => 'std',
	'Winkeys' => 'winkeys',
    },
    'ru' => {
	'Phonetic' => 'phonetic',
	'Tatar' => 'tt',
	'Typewriter' => 'typewriter',
	'Winkeys' => 'winkeys',
    },
    'se' => {
	'Dvorak' => 'dvorak',
	'Eliminate dead keys' => 'nodeadkeys',
	'Northern Saami' => 'smi',
	'Russian phonetic' => 'rus',
	'Russian phonetic, eliminate dead keys' => 'rus_nodeadkeys',
    },
    'si' => {
	'US keyboard with Slovenian digraphs' => 'unicodeus',
	'US keyboard with Slovenian letters' => 'us',
	'Use Slovenian digraphs' => 'unicode',
	'Use guillemots for quotes' => 'alternatequotes',
    },
    'sk' => {
	'Extended Backslash' => 'bksl',
	'qwerty' => 'qwerty',
	'qwerty, extended Backslash' => 'qwerty_bksl',
    },
    'sy' => {
	'Syriac' => 'syc',
	'Syriac phonetic' => 'syc_phonetic',
    },
    'th' => {
	'Pattachote' => 'pat',
	'TIS-820.2538' => 'tis',
    },
    'tr' => {
	'(F)' => 'f',
	'Alt-Q' => 'alt',
	'Kurdish, (F)' => 'ku_f',
	'Kurdish, Latin Alt-Q' => 'ku_alt',
	'Kurdish, Latin Q' => 'ku',
	'Sun dead keys' => 'sundeadkeys',
    },
    'ua' => {
	'Phonetic' => 'phonetic',
	'Standard RSTU' => 'rstu',
	'Standard RSTU on Russian layout' => 'rstu_ru',
	'Typewriter' => 'typewriter',
	'Winkeys' => 'winkeys',
    },
    'us' => {
	'Alternative international (former us_intl)' => 'alt-intl',
	'Dvorak' => 'dvorak',
	'International (with dead keys)' => 'intl',
	'Left handed Dvorak' => 'dvorak-l',
	'Right handed Dvorak' => 'dvorak-r',
	'Russian phonetic' => 'rus',
    },
);

1;
